gitlab/gitlab.go

```diff
@@ -90,6 +99,29 @@ func (hook Webhook) Parse(r *http.Request, events ...Event) (interface{}, error)
 
 	gitLabEvent := Event(event)
 
+	if gitLabEvent == SystemHooks {
```

Comment:
Can this block of logic be moved into the switch statement below?

Reply:
For this block we have 3 options:

1) **Leave it** - this block is present cause, as described in the first comment, the whole library checks for the HTTP-Header to determine which payload is sent and for gitlab system hooks it is always "System Hook" regardles of the payload content.
So what i have done here is if it is an `system hook` check the payload property `object_kind` and overwrite the `gitLabEvent` variable with the corresponding type. With that pre-processing I didn't have to change the switch below.

2) **Move it to the Switch below** - if the switch below should handle the system hook i have to do a switch in a switch stmnt cause the payload can be one of either push/tag/merge-request which would result in duplicated code.

3) **Rewrite the Event detection** - basically omit the header to determine the type of the payload and use the `object_kind` property in the json instead. This would lead to an additional json unmarshal for all events (which shouldn't be a problem at all imho)

Reply:
Hey @joeybloggs ,

I tried to implement the logic into the switch below and while doing it the code felt really awkward, cause of the multiple switch structures and the additional event allowed parsing.

In Fact if it is okay for you i would like to change the code to version 3 of the suggestions postet above, omitting the header and use the json property for all the detection work.

Here is a part of the code (not directly working):
```go

       ...rest of function...

        switch gitLabEvent {
	case BuildEvents:
		var pl BuildEventPayload
		err = json.Unmarshal([]byte(payload), &pl)
		return pl, err

	case SystemHookEvents:
		var sysPl SystemHookPayload
		err = json.Unmarshal([]byte(payload), &sysPl)
		if err != nil {
			return nil, err
		}
		switch sysPl.ObjectKind {
		case objectPush:
			gitLabEvent = PushEvents
		case objectTag:
			gitLabEvent = TagEvents
		case objectMergeRequest:
			gitLabEvent = MergeRequestEvents
		}
		err = checkEvents(events, gitLabEvent)
		if err != nil {
			return nil, err
		}
		switch gitLabEvent {
		case PushEvents:
			var pl PushEventPayload
			err = json.Unmarshal([]byte(payload), &pl)
			return pl, err

		case TagEvents:
			var pl TagEventPayload
			err = json.Unmarshal([]byte(payload), &pl)
			return pl, err

		case MergeRequestEvents:
			var pl MergeRequestEventPayload
			err = json.Unmarshal([]byte(payload), &pl)
			return pl, err
		}
	default:
		return nil, fmt.Errorf("unknown event %s", gitLabEvent)
	}
}

func checkEvents(events []Event, gitLabEvent Event) error {
	var found bool
	for _, evt := range events {
		if evt == gitLabEvent {
			return nil
		}
	}
	return ErrEventNotFound
}
```

Reply:
@Munsio If you could confirm one thing then I think I have a better way, do all `ObjectKind`'s within the `SystemHookEvents` map directly to another event?

eg. `objectPush` maps to `PushEvents`?

if so the switch could just be abstracted into a helper function to be called again within the `SystemHookEvents` case passing the correct event type along and then could just always listen for System Hooks and not have it as a configurable option.

Reply:
Okay i will give it a try to craft something 👍

Reply:
@joeybloggs hey i made an update.
I put the whole switch into an separate function and in there i check if the provided Event is allowed, than if it is an `SystemHook` i call it again with the correct event type.

Feedback appreciated!